test/modules/aws-security-group-integration.ts

```diff
@@ -131,36 +196,349 @@ describe('Security Group Integration Testing', () => {
+  );
+  it(
+    'deletes the source security group for non default',
+    query(`DELETE FROM security_group WHERE group_name = '${prefix}sgsourcetestnotdefault'`),
+  );
+  it(
+    'deletes the default security group for vpc',
+    query(
+      `DELETE FROM security_group WHERE vpc_id=(SELECT id FROM vpc WHERE cidr_block='192.${randIPBlock}.0.0/24')`,
+    ),
+  );
+
+  it(
+    'adds a new security group rule pointing to itself',
+    query(`
+  INSERT INTO security_group_rule(description, security_group_id, source_security_group, is_egress) 
+  VALUES ('${prefix}sgsourcetestrule', (SELECT id FROM security_group WHERE group_name='${prefix}sgforsource'),
+  (SELECT id FROM security_group WHERE group_name='${prefix}sgforsource'), false);
+  `),
+  );
+  it('creates the source security group rule pointing to itself', apply());
+
+  it(
+    'check security group rule expanded for TCP pointing to itself',
+    query(
+      `
+    SELECT *
+    FROM security_group_rule
+    WHERE source_security_group = (SELECT id FROM security_group WHERE group_name='${prefix}sgforsource') AND ip_protocol='tcp';
+  `,
+      (res: any[]) => expect(res.length).toBe(1),
+    ),
+  );
+
+  it(
+    'check security group rule expanded for UDP pointing to itself',
+    query(
+      `
+    SELECT *
+    FROM security_group_rule
+    WHERE source_security_group = (SELECT id FROM security_group WHERE group_name='${prefix}sgforsource') AND ip_protocol='udp';
+  `,
+      (res: any[]) => expect(res.length).toBe(1),
+    ),
+  );
+
+  it(
+    'check security group rule expanded for ICMP pointing to itself',
+    query(
+      `
+    SELECT *
+    FROM security_group_rule
+    WHERE source_security_group = (SELECT id FROM security_group WHERE group_name='${prefix}sgforsource') AND ip_protocol='icmp';
+  `,
+      (res: any[]) => expect(res.length).toBe(1),
+    ),
+  );
+
+  it(
+    'deletes the source security group rule pointing to itself',
+    query(
+      `DELETE FROM security_group_rule WHERE source_security_group = (SELECT id FROM security_group WHERE group_name='${prefix}sgforsource')`,
+    ),
+  );
+
+  it('applies the deletion of source security group rule', apply());
+
+  it(
+    'check no security group rules remain',
+    query(
+      `
+    SELECT *
+    FROM security_group_rule
+    WHERE source_security_group = (SELECT id FROM security_group WHERE group_name='${prefix}sgforsource');
+  `,
+      (res: any[]) => expect(res.length).toBe(0),
+    ),
+  );
+
+  it(
+    'deletes the original security group to test source',
+    query(`DELETE FROM security_group WHERE group_name = '${prefix}sgforsource'`),
+  );
+
+  it('applies the deletion of source security group rule', apply());
+
+  it(
+    'check no security group for source remain',
+    query(
+      `
+    SELECT *
+    FROM security_group
+    WHERE group_name = '${prefix}sgsourcetest' OR group_name='${prefix}sgforsource';
+  `,
+      (res: any[]) => expect(res.length).toBe(0),
+    ),
+  );
```

Comment:
I would like to propose one more set of security group tests. Often security group rules are written such that security group A is only accessible to security group B *and* vice versa, such that there is a recursive-looking intertwined situation between them.

It is possible to create these in the database by creating security group A, then security group B, then security group A's rule points at B and security group B's rule points at A, and to delete them in the reverse order, but making sure that the engine doesn't trip up on this is important.

Can we have a set of six `it` statements like this?

1. A query with four separate `INSERT`s to create these self-referential security groups.
2. An `apply` (confirming that it isn't tripped up by this structure)
3. A `SELECT` query to confirm these records do exist with `expect` assertions on it (likely something like `SELECT sg.id, sgr.source_security_group FROM security_group sg INNER JOIN security_group_rule sgr ON sg.id = sgr.security_group_id WHERE sgr.source_security_group IS NOT NULL`, assert that there should be two records, and that the `id` and `source_security_group` fields are opposite paired values between the two)
4. Four `DELETE` query to delete the security group rules *and* the security groups
5. Another `apply` (confirming that it can likewise drop these rules "atomically" and you don't have to delete one-at-a-time manually.
6. Another `SELECT` confirming that they are really gone.